Gce round 3

justnoise opened the pull request:
We've refactored the code to create, destroy and down instances, as well as create images for all cloud providers.  All the new code is hidden in `bookshelf/api_v3`.  The biggest change is moving to classes that represent cloud instances with a consistent interface (in `bookshelf/api_v3/cloud_instance.py`) across those classes.  

Each class has a constructor to create a cloud instance from a user-supplied configuration (`create_from_config`) or bring up a saved instance (running or not running) from a configuration and previously saved state information (`create_from_saved_state`).  The saved state data mostly just lists the instance name you've already provisioned.

Overall, when making changes to these classes, we found that propagating those changes across clouds was transparent and easy.  The classes should be able to support provisioning images for other purposes in the future.

sarum90 committed:
parametarize the network

sarum90 committed:
contd

Commit:
adding api_v3 and refactoring to classes representing each cloud provider

sarum90 committed:
Interface fix-ups

Commit:
better management of state

sarum90 committed:
add interface tests

sarum90 committed:
add provides documentation

Commit:
fixed up a few typos

Commit:
Merge remote-tracking branch 'origin/gce_round_3_aws_addon' into gce_round_3

sarum90 committed:
trying something

sarum90 committed:
ec2 implemented

Commit:
got rackspace and gce up and running correctly

sarum90 committed:
actually commit my work

sarum90 committed:
Merging

sarum90 committed:
actually test the current checkout

Commit:
fixed up gce and rackspace to do the right thing, GCE now uses service account credentials

Commit:
Merge branch 'gce_round_3' of github.com:ClusterHQ/bookshelf into gce_round_3

sarum90 committed:
Flatten configs for a nicer interface.

sarum90 committed:
resolve merge errors

sarum90 committed:
find ami from the correct location

Commit:
added additional auth methods for gce, fixed up errant field in rackspace

Commit:
Merge branch 'gce_round_3' of github.com:ClusterHQ/bookshelf into gce_round_3

sarum90 committed:
fixed some accidental changes

Commit:
use image_basename instead of description for naming slave images

Azulinho commented on .travis.yml:
```diff
@@ -14,11 +14,12 @@ matrix:
+  - TEST_SUITE=api_v3/test_interfaces.py
   # we can't run vagrant on Travis.CI, as it uses OpenVZ
   # so we need to skip the docker tests for now
   # - TEST_SUITE=test_docker.py
```
do you plan to introduce additional tests for the ec2, rackspace, gce calls ?
see bookshelf/tests/api_v2/test_rackspace.py and test_ec2.py

sarum90 replied:
Yup, will do in a follow up. Specifically https://github.com/ClusterHQ/bookshelf/pull/6 (FLOC-4156)

Azulinho commented on bookshelf/api_v3/cloud_instance.py:
```diff
@@ -0,0 +1,113 @@
+            to a given implementation
+        :param saved_state: The serialization state of an instance created by
+            serializing a previously
+
```
'a previously serialised instance'?

Azulinho commented on bookshelf/api_v3/ec2.py:
```diff
@@ -0,0 +1,321 @@
+
+
+class EC2State(PClass):
+    instance_id = field(type=unicode, mandatory=True, factory=unicode)
```
docstring please

Azulinho commented on bookshelf/api_v3/ec2.py:
```diff
@@ -0,0 +1,321 @@
+    """
+    Class representing an EC2 instance, that provides methods for interacting
+    with the instance.
+
```
maybe name it EC2instance ?
There are many types of objects in EC2 that are not just EC2 instances (ssh groups, key pairs, ELBs...) while bookshelf doesn't provide a method for managing those today (as it hasn't been needed so far). That is likely to happen in the future.
If/When that happens a class named EC2 would be imply too generic

justnoise replied:
Cool, makes sense to me!

Azulinho commented on bookshelf/api_v3/ec2.py:
```diff
@@ -0,0 +1,321 @@
+            credentials=parsed_config.credentials
+        )
+
+        timeout = 600
```
can we move timeout as a parameter and default it to 600 ?

sarum90 replied:
Sure, although we didn't bother adding it to the interface, so it only works for EC2 for now.

Azulinho commented on bookshelf/api_v3/gce.py:
```diff
@@ -0,0 +1,425 @@
+from cloud_instance import ICloudInstance, ICloudInstanceFactory, Distribution
+
+
+class GCEConfiguration(PClass):
```
docstring please

justnoise replied:
For the <cloud>Configuration and <cloud>State classes, would you like an explanation of each individual field (similar to what's in the CI-slave-images README)?

Azulinho replied:
Not really.
Only if you are really really bored.
If someone managed to get this far without reading the readme. Then they don't deserve any special assistance.

Azulinho commented on bookshelf/api_v3/gce.py:
```diff
@@ -0,0 +1,425 @@
+    instance_name = field(factory=unicode, mandatory=True)
+    ip_address = field(factory=unicode, mandatory=True)
+    distro = field(factory=unicode, mandatory=True)
+    zone = field(factory=unicode, mandatory=True)
```
does zone belong here or in    GCEConfiguration ?

justnoise replied:
We were shooting for configuration (the data in the yaml file) being a static file that doesn't change between provisioning runs.  The region can change based on what you'd like to do and the parameter is set when running `fab up` in on the command line.  Thus, it's not so muh provisioning configuration as a property of the instance you're provisioning.

Azulinho commented on bookshelf/api_v3/gce.py:
```diff
@@ -0,0 +1,425 @@
+
+    @classmethod
+    def create_from_config(cls, config, distro, region):
+        instance_name = "{}-{}".format(
```
docstring

justnoise replied:
The docstring will be the same as the docstring on the interface (`cloud_instance.ICloudInstanceFactory`). Lemmie know if that doesn't suffice.

Azulinho commented on bookshelf/api_v3/gce.py:
```diff
@@ -0,0 +1,425 @@
+        Shuts down the instance and creates and image from the disk.
+        Assumes that the disk name is the same as the instance_name (this is
+        the default behavior for boot disks on GCE).
+        """
```
is instance_name the equivalent of instance.id on ec2 ?

justnoise replied:
Correct. GCE uses names for most resources instead of IDs.

Azulinho commented on bookshelf/api_v3/gce.py:
```diff
@@ -0,0 +1,425 @@
+        ).execute())
+
+    def destroy(self):
+        log_yellow("downing server: {}".format(self.state.instance_name))
```
'downing' I like it

Azulinho commented on bookshelf/api_v3/rackspace.py:
```diff
@@ -0,0 +1,231 @@
+            region=region
+        )
+        instance = Rackspace(config, state)
+        instance.upload_key()
```
nice touch

Azulinho commented on bookshelf/api_v3/rackspace.py:
```diff
@@ -0,0 +1,231 @@
+        server = instance._ensure_instance_running()
+        # if we've restarted a terminated server, the ip address
+        # might have changed from our saved state, get the
+        # networking info and resave the state
```
is there value in doing this in Rackspace?
as starting a terminated server will always create a new server, since shutdown a box in rackspace always means destroy this box

justnoise replied:
Ah, I didn't know that. I'll test that out (just to make sure) and then get rid of the extra junk.

justnoise replied:
Cool.  Rackspace support for halting a server and bringing it back up sucks.  You have to reboot the thing to get it back up and running after halting it.  There's not much of a reason to down a server anyways (you still pay for the damn thing).  I'll remove downing a server and ensuring it's running from the RackspaceInstance.

Azulinho commented on bookshelf/tests/api_v3/test_interfaces.py:
```diff
@@ -0,0 +1,43 @@
+
+
+if __name__ == '__main__':
+    unittest.main(verbosity=4, failfast=True)
```
do you plan to add additional tests here?
see:https://github.com/ClusterHQ/bookshelf/blob/master/bookshelf/tests/api_v2/test_rackspace.py
and
https://github.com/ClusterHQ/bookshelf/blob/master/bookshelf/tests/api_v2/test_ec2.py

Azulinho commented on setup.py:
```diff
@@ -4,15 +4,17 @@
-     )
+setup(
+    name='Bookshelf',
+    version='2.0.4',
```
time to bump that up
3.0.0

Azulinho commented on setup.py:
```diff
@@ -4,15 +4,17 @@
+    url='https://www.github.com/ClusterHQ/Bookshelf/',
+    packages=['bookshelf', 'bookshelf.api_v2', 'bookshelf.tests',
+              'bookshelf.tests.api_v2', ],
+    install_requires=['cuisine', 'fabric', 'pyrax', 'boto',
```
thinks .api_v3 is missing here

Azulinho commented:
This is awesome,
Thanks a lot for this.

Can you do a couple of things?
update the version to 3.0.0 in setup.py
run python setup.py sdist
and add that .tar.gz to the repo as part of this branch
(this is not being published to pypi yet)

merge it
and feel free to add some additional tests in a separate PR

sarum90 committed:
Added tests to verify the functionality of the ICloudInstance providers.

Decided to write a single test case that can be executed on each of the
available ICloudInstance providers.

Note that presently these require you to delete the image they create manually
(WIP).

Commit:
added image listing and deletion, and incorporated reviewer feedback from PR

Commit:
Merge branch 'gce_round_3' of github.com:ClusterHQ/bookshelf into gce_round_3

sarum90 committed:
update the tests

sarum90 committed:
update test to use the new interface

Commit:
statusMessage isnt available, just return status from delete_image

Commit:
Merge branch 'gce_round_3' of github.com:ClusterHQ/bookshelf into gce_round_3

sarum90 commented:
Closing this in favor of https://github.com/ClusterHQ/bookshelf/pull/5, which is effectively a sqashing of this.
